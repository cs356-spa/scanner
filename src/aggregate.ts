import { writeFileSync } from "fs";
import { readFileSync } from "fs";

const domains = JSON.parse(readFileSync(__dirname + "/output.json").toString());

const domainsNoneDetected = domains.filter(e => Object.keys(e.output).length === 0);
const domainsDetected = domains.filter(e => Object.keys(e.output).length > 0);

const packages = {};
const packageTotals = {};

for (let domain of domainsDetected) {
  for (let packageName in domain.output) {
    const packageDetails = domain.output[packageName];
    if (!packages[packageName]) packages[packageName] = {};
    if (!packageTotals[packageName]) packageTotals[packageName] = 0;
    if (!packages[packageName][packageDetails.version]) packages[packageName][packageDetails.version] = 0;
    packages[packageName][packageDetails.version]++;

    packageTotals[packageName]++;
  }
}

// const stats = `
// Total domains: ${domains.length}
// Domains without versions: ${domainsNoneDetected.length}
// Domains with versions: ${domainsDetected.length}
// ${JSON.stringify(packageTotals, null, 2)}
// ${JSON.stringify(packages, null, 2)}
// `;

const stats = {
  totalDomains: domains.length,
  domainsNoneDetected: domainsNoneDetected.length,
  domainsDetected: domainsDetected.length,
  packageTotals,
  packages
};

writeFileSync(__dirname + "/stats.json", JSON.stringify(stats, null, 2));


packages["react"].total = Object.values(packages["react"]).reduce((a: number, b: number) => a + b, 0);
const reactVersions = Object.keys(packages["react"]).map(e => ({version: e, count: packages['react'][e]}) ).sort((a, b) => b.count - a.count);
console.log(reactVersions);

const latex = `\\begin{tabular}{ll}
\\hline
\\multicolumn{1}{|l|}{Package name} & \\multicolumn{1}{l|}{Total} \\\\ \\hline
${reactVersions.map(e => `\\multicolumn{1}{|l|}{${e.version}} & \\multicolumn{1}{l|}{${e.count}} \\\\ \\hline`).join("\n")}
\\end{tabular}
`;

// console.log(latex);