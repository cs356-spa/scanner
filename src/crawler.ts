import * as fs from "fs";
import { Cluster } from 'puppeteer-cluster';
import { staticScanForSPAFramework, mergeDetectorOutput, SpaDetectorOutput, isBundledFileStrict }  from './static_detector';
import { parseCSV } from "./top-sites";

const { dynamicScanForSPAFramework } = require("./dynamic_detector");
const debug = require("debug")("spa");

const sleep = (ms: number) => new Promise(res => setTimeout(res, ms));

const DEBUG_DISPLAY_NUM_SITES_WITH_BUNDLE = true;
let numSitesWithBundle = 0;

async function runOnPage(page, domain) {
  const sources: {url: string, content: string}[] = [];
  page.on("console", msg => {
    // TODO: msg.args() to capture console content
  });
  // page.evaluateOnNewDocument(() => {
  //   // TODO: add code that will be run in the browser context
  // });

  await page.setRequestInterception(true);
  page.on("request", async request => {
    try {
      if (request.isNavigationRequest() && request.redirectChain().length > 10) { // prevent too many redirects failure.
        await request.abort();
      } else {
        await request.continue();
      }
    } catch {}
  });

  // await page.setRequestInterception(true); // Used to inject custom scripts
  page.on("response", async response => {
    try {
      const content = await response.text(); // .toLowerCase(); // Don't do this!!!
      sources.push({ url: response.url(), content })
    } catch {
      // console.log(`x Response read failed for ${response.url()}`);
    }
  });

  debug(`Navigating to ${domain}`);
  // Racing between networkidle2 and a given timeout. If wait for TOO long, page explodes under puppeteer...
  await Promise.race([page.goto(domain, {waitUntil: "networkidle2"}), sleep(10000)])
  // await page.goto(domain, {waitUntil: "networkidle2"});

  // IMPORTANT: we need to add the HTML after rendering into sources for search (in the Angular case)
  try {
    // TODO: mysterious error keeps popping up:
    // Error: Execution context was destroyed, most likely because of a navigation.
      // at rewriteError (/Users/kun/Desktop/Fall_2020/CS356/Project/scanner/node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:261:23)
      // at runMicrotasks (<anonymous>)
      // at processTicksAndRejections (internal/process/task_queues.js:93:5)
      // at ExecutionContext._evaluateInternal (/Users/kun/Desktop/Fall_2020/CS356/Project/scanner/node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:215:61)
      // at ExecutionContext.evaluate (/Users/kun/Desktop/Fall_2020/CS356/Project/scanner/node_modules/puppeteer/lib/cjs/puppeteer/common/ExecutionContext.js:106:16)
      // at DOMWorld.content (/Users/kun/Desktop/Fall_2020/CS356/Project/scanner/node_modules/puppeteer/lib/cjs/puppeteer/common/DOMWorld.js:115:16)
      // at Page.content (/Users/kun/Desktop/Fall_2020/CS356/Project/scanner/node_modules/puppeteer/lib/cjs/puppeteer/common/Page.js:819:16)
    // const finalPageContent = await page.content();
    // console.log(await page.evaluate(function () { console.log("HI"); return window.location.href }));
    // const finalPageContent = await page.evaluate(function () { return document.documentElement.outerHTML; });
    const finalPageContent = await page.content();
    sources.push({ url: domain, content: finalPageContent });
  } catch (_e) {
    console.log(`>>> Content Error for ${domain}`);
  }
  // TODO: consider forcefully push LICENSE file into list of sources. This requires us to somehow be able to identify URLs to LICENSE files
  // For example, Airbnb will inline URL to LICENSE file: https://a0.muscache.com/airbnb/static/packages/moment-3cf2a832.js, pointing to https://a0.muscache.com/airbnb/static/packages/moment-879e3275.js.LICENSE.txt

  const staticOutput = staticScanForSPAFramework(sources, new URL(domain).hostname);
  // Catch the potential case where dynamic evaluation fails.
  const dynamicOutput = await page.evaluate(dynamicScanForSPAFramework).catch(() => ({}));

  if (DEBUG_DISPLAY_NUM_SITES_WITH_BUNDLE) {
    const hasBundledFile = sources.reduce((acc, s) => acc || isBundledFileStrict(s), false);
    if (hasBundledFile) {
      numSitesWithBundle++;
    }
  }

  debug(`Finish task for ${domain}`);
  const output = mergeDetectorOutput(staticOutput, dynamicOutput);

  try {
    // Kill the page when necessary. This is required to allow a clear profile when navigating to the next page.
    // page.evaluate also has the change of failure similar to above.
    await page.evaluate(() => window.stop());
  } catch (_e) {}

  return output;
}

async function createCustomBrowser() {
  const customArgs = [
    '--disable-dev-shm-usage',
    '--no-sandbox',
    // `--load-extension=${path.resolve("./extensions/react_devtools/")}`
  ];
  return {
    // defaultViewport: null,
    // executablePath: process.env.chrome,
    headless: true, // WARNING: we have to make it NOT headless to get the extension to work! This is a sad compromise... See https://bugs.chromium.org/p/chromium/issues/detail?id=706008#c5
    // ignoreDefaultArgs: ["--disable-extensions"],
    args: customArgs,
    product: 'chrome',
    executablePath: 'google-chrome-stable'
  };
}

export async function main(domains: string[]) {
  const browserArgs = await createCustomBrowser();
  const cluster = await Cluster.launch({
    concurrency: Cluster.CONCURRENCY_CONTEXT,
    maxConcurrency: 50,
    puppeteerOptions: browserArgs
  });
  console.log('starting browser...');

  const results: {domain: string, output: SpaDetectorOutput}[] = [];
  for (let i in domains) {
    const domain = domains[i];
    cluster.queue(async ({ page }) => {
      try {
        results.push({
          domain,
          output: await runOnPage(page, domain)
        });
        console.log(domain, i, domains.length);
      } catch (e) {
        console.error(`>>> on domain "${domain}"`, e);
      }
    });
  };

  await cluster.idle();
  debug("Shutting down cluster...")
  await cluster.close();
  debug("Done");

  if (DEBUG_DISPLAY_NUM_SITES_WITH_BUNDLE) {
    console.log(`>>> ${numSitesWithBundle}/${domains.length} sites has bundled file`);
  }

  return results.sort((a, b) => a.domain.localeCompare(b.domain));
}

/**
 * Using this comparison, we can use this file both as a module AND as a running file.
 * Usage: ts-node ./crawler.ts ./example/top_100.csv
 */
if (require.main === module) {
  if (process.argv.length < 3) {
    console.error("Error: please provide filename of domains");
  }
  const domainCSVFilename = process.argv[2];
  const domainCSVRawContent = fs.readFileSync(domainCSVFilename, "utf8");
  const domains = parseCSV(domainCSVRawContent);

  // TODO: implement read csv file. need PR master merge.
  const result = main(domains.map(d => `http://${d.Domain}`));
  result.then(res => {
    console.log(`>>> In total ${res.length} sites successfully scanned to some extent`);
  })
}