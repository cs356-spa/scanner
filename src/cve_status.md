### Some interesting CVEs of libraries we can briefly discuss on

+ React
  + CVE-2018-6341: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-6341
    + React applications which rendered to HTML using the ReactDOMServer API were not escaping user-supplied attribute names at render-time. That lack of escaping could lead to a cross-site scripting vulnerability. This issue affected minor releases 16.0.x, 16.1.x, 16.2.x, 16.3.x, and 16.4.x. It was fixed in 16.0.1, 16.1.2, 16.2.1, 16.3.3, and 16.4.2.
    + Blog post: https://reactjs.org/blog/2018/08/01/react-v-16-4-2.html
      + Vue and Svelte might have similar issues: https://github.com/vuejs/vue/commit/c28f79290d57240c607d8cec3b3413b49702e1fb
        + Vue coordinated tweet: https://twitter.com/vuejs/status/1024754536877973504
    + From sample run, there are actually sites using 16.2.0, which is a vulnerable version
      + Would be interesting to know which site is like this.
+ Angular.js
  + CVE-2020-7676: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-7676
    + angular.js prior to 1.8.0 allows cross site scripting. The regex-based input HTML replacement may turn sanitized code into unsanitized one. Wrapping `<option>` elements in `<select>` ones changes parsing behavior, leading to possibly unsanitizing code.
    + Github fix: https://github.com/angular/angular.js/pull/17028
  + CVE-2019-14863: https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2019-14863
    + There is a vulnerability in all angular versions before 1.5.0-beta.0, where after escaping the context of the web application, the web application delivers data to its users along with other trusted dynamic content, without validating it.
    + Snyk report: https://snyk.io/vuln/npm:angular:20150807
      + Github fix: https://github.com/angular/angular.js/pull/12524


+ Samples from other packages
  + Moment.js
    + DoS exploit vector:
      + Before 2.15.2
      + Why: https://gist.github.com/grnd/50192ce22681848a7de812d95241b7fc
      + Fix: https://github.com/moment/moment/commit/663f33e333212b3800b63592cd8e237ac8fabdb9



+ Other useful vulnerability DB
  + Snyk: https://snyk.io/vuln