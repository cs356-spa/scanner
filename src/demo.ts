import * as path from 'path';
import * as puppeteer from 'puppeteer';
import { Cluster } from 'puppeteer-cluster';
import { staticScanForSPAFramework, mergeDetectorOutput }  from './static_detector';

const { dynamicScanForSPAFramework } = require("./dynamic_detector");
const debug = require("debug")("spa");

/**
 * Async wrapper to sleep for a while. Try `await sleep(1000)` to sleep for 1 second.
 * @param {number} ms Milliseconds to sleep.
 * @return {Promise<void>}
 */
function sleep(ms) {
  return new Promise(res => setTimeout(res, ms));
}

async function runOnPage(page, domain) {
  const sources: {url: string, content: string}[] = [];
  page.on("console", msg => {
    // TODO: msg.args() to capture console content
  });
  page.evaluateOnNewDocument(() => {
    // TODO: add code that will be run in the browser context
  });

  await page.setRequestInterception(true);
  page.on("request", async request => {
    try {
      if (request.isNavigationRequest() && request.redirectChain().length > 10) { // prevent too many redirects failure.
        await request.abort();
      } else {
        await request.continue();
      }
    } catch {}
  });

  // await page.setRequestInterception(true); // Used to inject custom scripts
  page.on("response", async response => {
    try {
      const content = (await response.text()).toLowerCase();
      sources.push({ url: response.url(), content })
    } catch {
      // console.log(`x Response read failed for ${response.url()}`);
    }
  });

  debug(`Navigating to ${domain}`);
  await page.goto(domain, {waitUntil: "networkidle2"});

  // IMPORTANT: we need to add the HTML after rendering into sources for search (in the Angular case)
  const finalPageContent = await page.content();
  sources.push({ url: domain, content: finalPageContent });
  // TODO: consider forcefully push LICENSE file into list of sources. This requires us to somehow be able to identify URLs to LICENSE files
  // For example, Airbnb will inline URL to LICENSE file: https://a0.muscache.com/airbnb/static/packages/moment-3cf2a832.js, pointing to https://a0.muscache.com/airbnb/static/packages/moment-879e3275.js.LICENSE.txt

  const staticOutput = staticScanForSPAFramework(sources, new URL(domain).hostname);
  const dynamicOutput = await page.evaluate(dynamicScanForSPAFramework);

  console.log(mergeDetectorOutput(staticOutput, dynamicOutput));
  debug(`Finish task for ${domain}`);
}

const testDomains = [
  // Angular test targets
  "https://angular.io/",
  // "https://www.delta.com/", // Delta seems to have protection against crawler, causing an infinite redirect to self.
  "https://clarity.design/",
  "https://ng.ant.design/docs/introduce/en",

  // Vue test targets
  "https://vuejs.org/",

  // React test targets
  // Dynamic extension based scan give concrete results (when NOT under headless, and extension is loaded)
  "https://reactjs.org",
  "https://airbnb.com",
  "https://webpack.js.org",
  // (Occasionally, sometimes not working) Only detecting React presence (old versions of React, likely before 0.15.x/15.x)
  "https://facebook.com",

  // Ember test targets
  "https://emberjs.com"
]

async function createCustomBrowser() {
  const customArgs = [
    `--load-extension=${path.resolve("./extensions/react_devtools/")}`
  ];
  return {
    // defaultViewport: null,
    // executablePath: process.env.chrome,
    // headless: false, // WARNING: we have to make it NOT headless to get the extension to work! This is a sad compromise... See https://bugs.chromium.org/p/chromium/issues/detail?id=706008#c5
    ignoreDefaultArgs: ["--disable-extensions"],
    args: customArgs,
  };
}

export async function main() {
  const browserArgs = await createCustomBrowser();
  const cluster = await Cluster.launch({
    concurrency: Cluster.CONCURRENCY_CONTEXT,
    maxConcurrency: 50,
    puppeteerOptions: browserArgs
  });
  console.log('starting browser...');

  for (let domain of testDomains) {
    cluster.queue(async ({ page }) => {
      try {
        await runOnPage(page, domain);
      } catch (e) {
        console.error(e);
      }
    });
  };

  await cluster.idle();
  debug("Shutting down cluster...")
  await cluster.close();
  debug("Done");
}