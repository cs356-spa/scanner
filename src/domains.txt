# Angular test targets
https://angular.io/
# "https://www.delta.com/", # Delta seems to have protection against crawler, causing an infinite redirect to self
https://clarity.design/
https://ng.ant.design/docs/introduce/en

# Vue test targets
https://vuejs.org/

# React test targets
# Dynamic extension based scan give concrete results (when NOT under headless, and extension is loaded)
https://reactjs.org
https://airbnb.com
https://webpack.js.org
# (Occasionally, sometimes not working) Only detecting React presence (old versions of React, likely before 0.15.x/15.x)
https://facebook.com

# Ember test targets
https://emberjs.com