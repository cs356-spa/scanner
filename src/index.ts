import { download } from "./top-sites";

(async () => {
  const sites = await download();
  console.log(sites);
})();