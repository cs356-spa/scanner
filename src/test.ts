const path = require('path');

const puppeteer = require("puppeteer");
const { staticScanForSPAFramework, mergeDetectorOutput } = require("./static_detector");
const { dynamicScanForSPAFramework } = require("./dynamic_detector");
const debug = require("debug")("spa");

/**
 * Async wrapper to sleep for a while. Try `await sleep(1000)` to sleep for 1 second.
 * @param {number} ms Milliseconds to sleep.
 * @return {Promise<void>}
 */
function sleep(ms) {
  return new Promise(res => setTimeout(res, ms));
}

async function runOnPage(page, domain) {
  const sources: {url: string, content: string}[] = [];
  page.on("console", msg => {
    // TODO: msg.args() to capture console content
  });
  page.evaluateOnNewDocument(() => {
    // TODO: add code that will be run in the browser context
  });

  await page.setRequestInterception(true);
  page.on("request", async request => {
    try {
      if (request.isNavigationRequest() && request.redirectChain().length > 10) { // prevent too many redirects failure.
        await request.abort();
      } else {
        await request.continue();
      }
    } catch {}
  });

  // await page.setRequestInterception(true); // Used to inject custom scripts
  page.on("response", async response => {
    try {
      const content = (await response.text()).toLowerCase();
      sources.push({ url: response.url(), content })
    } catch {
      // console.log(`x Response read failed for ${response.url()}`);
    }
  });

  debug(`Navigating to ${domain}`);
  await page.goto(domain, {waitUntil: "networkidle2"});

  // IMPORTANT: we need to add the HTML after rendering into sources for search (in the Angular case)
  const finalPageContent = await page.content();
  sources.push({ url: domain, content: finalPageContent });
  // TODO: consider forcefully push LICENSE file into list of sources. This requires us to somehow be able to identify URLs to LICENSE files
  // For example, Airbnb will inline URL to LICENSE file: https://a0.muscache.com/airbnb/static/packages/moment-3cf2a832.js, pointing to https://a0.muscache.com/airbnb/static/packages/moment-879e3275.js.LICENSE.txt

  const staticOutput = staticScanForSPAFramework(sources, new URL(domain).hostname);
  const dynamicOutput = await page.evaluate(dynamicScanForSPAFramework);

  console.log(mergeDetectorOutput(staticOutput, dynamicOutput));
  debug(`Finish task for ${domain}`);
}

const testDomains = [
  // Angular test targets
  "https://angular.io/",
  // "https://www.delta.com/", // Delta seems to have protection against crawler, causing an infinite redirect to self.
  "https://clarity.design/",
  "https://ng.ant.design/docs/introduce/en",

  // Vue test targets
  "https://vuejs.org/",

  // React test targets
  // Dynamic extension based scan give concrete results (when NOT under headless, and extension is loaded)
  "https://reactjs.org",
  "https://airbnb.com",
  "https://webpack.js.org",
  // (Occasionally, sometimes not working) Only detecting React presence (old versions of React, likely before 0.15.x/15.x)
  "https://facebook.com",

  // Ember test targets
  "https://emberjs.com"
]

async function createCustomBrowser() {
  const customArgs = [
    `--load-extension=${path.resolve("./extensions/react_devtools/")}`
  ];
  const browser = await puppeteer.launch({
    // defaultViewport: null,
    // executablePath: process.env.chrome,
    // headless: false, // WARNING: we have to make it NOT headless to get the extension to work! This is a sad compromise... See https://bugs.chromium.org/p/chromium/issues/detail?id=706008#c5
    ignoreDefaultArgs: ["--disable-extensions"],
    args: customArgs,
  });
  return browser;
}

export async function main() {
  // const browser = await puppeteer.launch();
  const browser = await createCustomBrowser();
  const page = await browser.newPage();
  for (const domain of testDomains) {
    try {
      await runOnPage(page, domain);
    } catch (e) {
      console.log(e);
    }
  }
  debug("Shutting down browser...")
  await browser.close();
}