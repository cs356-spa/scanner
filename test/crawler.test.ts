import { main } from '../src/crawler';

describe('crawler', () => {
  test('basic list of domains', async () => {
    const domains = [
      "https://angular.io/",
      "https://vuejs.org/",
      "https://reactjs.org",
      "https://airbnb.com",
      "https://webpack.js.org",
      "https://facebook.com",
      "https://emberjs.com",
    ];
    const results = await main(domains);
    expect(results).toMatchSnapshot();
  })
})