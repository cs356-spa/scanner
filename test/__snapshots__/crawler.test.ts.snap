// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crawler basic list of domains 1`] = `
Array [
  Object {
    "domain": "https://airbnb.com",
    "output": Object {
      "react": Object {
        "confidence": 1,
        "isStatic": true,
        "reasonURL": "https://a0.muscache.com/airbnb/static/packages/react-0e221225.js",
        "version": "16.13.1",
      },
    },
  },
  Object {
    "domain": "https://angular.io/",
    "output": Object {
      "angular": Object {
        "confidence": 2,
        "isStatic": true,
        "reasonURL": "https://angular.io/",
        "version": "10.1.3",
      },
    },
  },
  Object {
    "domain": "https://emberjs.com",
    "output": Object {
      "ember": Object {
        "confidence": 2,
        "isStatic": false,
        "reasonURL": "https://emberjs.com/",
        "version": "3.16.3",
      },
    },
  },
  Object {
    "domain": "https://facebook.com",
    "output": Object {
      "react": Object {
        "confidence": 1,
        "isStatic": true,
        "reasonURL": "https://static.xx.fbcdn.net/rsrc.php/v3i7M54/yW/l/en_US/Sc3qJZQohzL.js?_nc_x=Ij3Wp8lg5Kz",
        "version": "17.0.0-alpha.0-d95c4938d",
      },
    },
  },
  Object {
    "domain": "https://reactjs.org",
    "output": Object {
      "react": Object {
        "confidence": 1,
        "isStatic": true,
        "reasonURL": "https://reactjs.org/framework-5096ca5d206149498093.js",
        "version": "17.0.0-rc.3",
      },
    },
  },
  Object {
    "domain": "https://vuejs.org/",
    "output": Object {
      "vue": Object {
        "confidence": 2,
        "isStatic": false,
        "reasonURL": "https://vuejs.org/",
        "version": "2.6.12",
      },
    },
  },
  Object {
    "domain": "https://webpack.js.org",
    "output": Object {
      "react": Object {
        "confidence": 1,
        "isStatic": true,
        "reasonURL": "https://webpack.js.org/vendor.bundle.js",
        "version": "16.13.1",
      },
    },
  },
]
`;
